backend/elasticsearch/src/main/java/org/hibernate/search/backend/elasticsearch/search/predicate/impl/AbstractElasticsearchSearchNestedPredicateBuilder.java

```diff
@@ -0,0 +1,58 @@
+/*
+ * Hibernate Search, full-text search for your domain model
+ *
+ * License: GNU Lesser General Public License (LGPL), version 2.1 or later
+ * See the lgpl.txt file in the root directory or <http://www.gnu.org/licenses/lgpl-2.1.html>.
+ */
+package org.hibernate.search.backend.elasticsearch.search.predicate.impl;
+
+import java.util.List;
+
+import org.hibernate.search.backend.elasticsearch.gson.impl.JsonAccessor;
+
+import com.google.gson.JsonObject;
+
+public abstract class AbstractElasticsearchSearchNestedPredicateBuilder extends AbstractElasticsearchSearchPredicateBuilder {
+
+	private static final JsonAccessor<String> PATH_ACCESSOR = JsonAccessor.root().property( "path" ).asString();
+	private static final JsonAccessor<JsonObject> QUERY_ACCESSOR = JsonAccessor.root().property( "query" ).asObject();
+
+	private final List<String> nestedPathHierarchy;
+
+	public AbstractElasticsearchSearchNestedPredicateBuilder(List<String> nestedPathHierarchy) {
+		this.nestedPathHierarchy = nestedPathHierarchy;
+	}
+
+	@Override
+	public final JsonObject build(ElasticsearchSearchPredicateContext context) {
+		JsonObject result = super.build( context );
+		if ( nestedPathHierarchy != null && !nestedPathHierarchy.isEmpty() ) {
+			result = applyImplicitNested( result, nestedPathHierarchy, context );
+		}
+		return result;
+	}
+
+	public static JsonObject applyImplicitNested(JsonObject partialResult, List<String> nestedPathHierarchy, ElasticsearchSearchPredicateContext context) {
+		JsonObject result = partialResult;
+
+		// traversing the nestedPathHierarchy in the inverted order
+		for ( int i = 0; i < nestedPathHierarchy.size(); i++ ) {
+			String path = nestedPathHierarchy.get( nestedPathHierarchy.size() - 1 - i );
+			if ( context.isExplicitNested( path ) ) {
```

Comment:
Knowing that the nested path hiarchy, in reverse order, will look like this:

* `nested1.nested2.nested3`
* `nested1.nested2`
* `nested1`

... and that `isExplicitNested()` checks `theExplicitPath.startsWith(theHierarchyComponent)`.

You will have something like that:

* First iteration: `"nested1".startsWith("nested1.nested2.nested3")` => `false`
* Second iteration: `"nested1".startsWith("nested1.nested2")` => `false`
* Second iteration: `"nested1".startsWith("nested1")` => `true`

It seems to me that you could just make `isExplicitNested()` check `theExplicitPath.equals(theHierarchyComponent)`. From what I understand, this will be strictly equivalent, and simpler.

Better yet, you could make the code more similar to the Lucene backend: remove `isExplicitNested()` and just write `context.getNestedPath().equals( path )` here.

Reply:
Yes, we got this additional benefit by processing in reverse order. I will add the test case you described above.